overlord/devicestate/devicestate.go

```diff
@@ -410,32 +409,40 @@ func Remodel(st *state.State, new *asserts.Model) ([]*state.TaskSet, error) {
 	var lastDownloadInChain, firstInstallInChain *state.Task
 	var prevDownload, prevInstall *state.Task
 	for _, ts := range tss {
-		// make sure all things happen sequentially:
+		// make sure all things happen sequentially
+		// Terminology
+		// A <- B means B waits for A
+		// "download,verify" are part of the "Download" phase
+		// "link,start" is part of "Install" phase
+		//
 		// - all tasks inside ts{Download,Install} already wait for
-		//   each other
-		//   Out chains look like this:
-		//     install1 <- verify1 <- download1
-		//     install2 <- verify2 <- download2
+		//   each other so the chains look something like this:
+		//     download1 <- verify1 <- install1
+		//     download2 <- verify2 <- install2
+		//     download3 <- verify3 <- install3
 		// - add wait of each first ts{Download,Install} task for
 		//   the last previous ts{Download,Install} task
 		//   Our chains now looks like:
-		//     download1 (no waits)
-		//     download2 <- verify1
-		//     install1 <- verify1 <- download1
-		//     install2 <- verify2 <- install1
-		firstDownload, lastDownload, firstInstall, lastInstall := extractDownloadInstallEdgesFromTs(ts)
+		//     download1 <- verify1 <- install1 (as before)
+		//     download2 <- verify2 <- install2 (as before)
+		//     download3 <- verify3 <- install3 (as before)
+		//     verify1 <- download2 (added)
+		//     verify2 <- download3 (added)
+		//     install1  <- install2 (added)
+		//     install2  <- install3 (added)
+		downloadStart, downloadLast, installFirst, installLast := extractDownloadInstallEdgesFromTs(ts)
 		if prevDownload != nil {
-			firstDownload.WaitFor(prevDownload)
+			downloadStart.WaitFor(prevDownload)
```

Comment:
mostly a remark, but we don't strictly need to serialize the download, just the install phases,  we could have the first install wait for all the downloadLasts

Reply:
While I agree with Samuele on this I would remark that unless we do something smart with network congestion (e.g. serialize downloads elsewhere) this will have the positive effect of working effectively on slower links that are easily congested by parallel downloads. In some cases downloading multiple files in parallel will cause downloads to fail whereas the same set processed sequentially would work.

Reply:
it probably makes sense at some point to introduce some kind of max-parallel-downloads option

Reply:
I add a comment for now, serializing it is nice for testing but I agree its not really needed.